core/src/main/scala/dev/profunktor/pulsar/Producer.scala

```diff
@@ -126,30 +138,75 @@ object Producer {
   /**
     * It creates a simple [[Producer]] with the supplied settings.
     */
-  private def make[F[_]: FutureLift: Parallel: Sync, E](
+  private def make[F[_]: FutureLift: Parallel: Sync, E: SeqIdMaker](
       client: Pulsar.T,
       topic: Topic.Single,
       settings: Settings[F, E]
   ): Resource[F, Producer[F, E]] = {
-    def configure[A](
-        producerBuilder: ProducerBuilder[A]
+    def batchingConf[A](
+        builder: ProducerBuilder[A]
     ): ProducerBuilder[A] =
       settings.batching match {
+        case Batching.Disabled =>
+          builder.enableBatching(false)
         case Batching.Enabled(delay, _) =>
-          producerBuilder
+          builder
             .enableBatching(true)
-            .batchingMaxPublishDelay(
-              delay.toMillis,
-              TimeUnit.MILLISECONDS
-            )
+            .batchingMaxPublishDelay(delay.toMillis, TimeUnit.MILLISECONDS)
             .batchingMaxMessages(5)
-            .topic(topic.url.value)
-        case Batching.Disabled =>
-          producerBuilder
-            .enableBatching(false)
-            .topic(topic.url.value)
       }
 
+    //https://pulsar.apache.org/docs/en/cookbooks-deduplication/
+    def deduplicationConf[A](
+        builder: ProducerBuilder[A]
+    ): ProducerBuilder[A] =
+      settings.deduplication match {
+        case Deduplication.Disabled => builder
+        case Deduplication.Enabled  => builder.sendTimeout(0, TimeUnit.SECONDS)
+      }
+
+    def configure[A]: ProducerBuilder[A] => ProducerBuilder[A] =
+      b => deduplicationConf(batchingConf(b)).topic(topic.url.value)
+
+    /*
+     * Internal function that manages Sequence IDs whenever deduplication is enabled.
+     * Otherwise, it defaults to the underlying default Java implementation.
+     *
+     * The acquisition of a new `sequenceId` is dictated by the [[SeqIdMaker]] typeclass.
+     */
+    def sendMessage[A](
+        p: JProducer[A],
+        msg: E,
+        enc: E => A,
+        key: MessageKey,
+        properties: Map[String, String],
+        prevMsgs: Ref[F, Option[E]]
+    ): F[MessageId] = {
+      val _msg = enc(msg)
+
+      def cont(f: TypedMessageBuilder[A] => TypedMessageBuilder[A]) =
+        settings.logger(msg)(topic.url) &>
+            FutureLift[F].futureLift {
+              f(
+                p.newMessage()
+                  .value(_msg)
+                  .properties(properties.asJava)
+                  .withShardKey(settings.shardKey(msg))
+                  .withMessageKey(key)
+              ).sendAsync()
+            }
+
+      settings.deduplication match {
+        case Deduplication.Enabled =>
+          prevMsgs.modify { prev =>
+            val nextId = SeqIdMaker[E].next(p.getLastSequenceId(), prev, msg)
+            Some(msg) -> cont(_.sequenceId(nextId))
+          }.flatten
```

Comment:
I'm suspicious about this - I wonder if the test failure is a race condition here

Can you make it stop flaking if you slap a semaphore(1) on this section

Reply:
Good eye! I can actually move the `cont` effect outside the `modify` block and see if that helps...

Reply:
Changed it to this in #20, no luck

```scala
prevMsgs
  .modify { prev =>
    val nextId = seqIdMaker
      .asInstanceOf[SeqIdMaker[E]]
      .next(p.getLastSequenceId(), prev, msg)
    Some(msg) -> nextId
  }
  .flatMap(nextId => cont(_.sequenceId(nextId)))
```

I'm going to add a log for that `nextId` to see if it's the right one

Reply:
I don't think changing `flatten` to `flatMap` changes the actual behavior *at all* though - it just changes how the method is structured.

In both cases, we have `modify state >> async interruption point >> execute cont`. I know it's probably not reasonable to have the final production code using a lock here, but I'd still check with a Semaphore

Reply:
Why are concurrency problems never solved by the easy thing? :sigh: